### PR TITLE
Add directlyEmbedCarthageDependencies to Target

### LINKED ISSUE #359: Add ability to embed Carthage frameworks without using `copy-frameworks`
On our project we have many test targets, and all of them depend on Carthage versions of Quick and Nimble. When we build this with parallel builds enabled a Carthage bug can be hit where two instances of `copy-frameworks` try to strip the same dSYM (https://github.com/Carthage/Carthage/issues/2531). Since test targets don't really need to strip anything at all, it would be good to be able to specify on a target that we want to embed the Carthage frameworks instead of using `copy-frameworks`. It can default properly for test targets.

I will add a `directlyEmbedCarthageDependencies` to `Target`. Right now it would be `true` for `.macOS` platform, and I'll extend it to be `true` for test targets as well.

### PR DESCRIPTION
**Pull request by brentleyjones**
Resolves #359.

### CODE REVIEW

File: Sources/XcodeGenKit/PBXProjGenerator.swift [hunk 1]
```diff
@@ -554,12 +557,32 @@ public class PBXProjGenerator {
...
+        for dependency in carthageDependencies {
+            guard target.type != .staticLibrary else { break }
+            
+            let embed = dependency.embed ?? target.shouldEmbedDependencies
+            
+            var platformPath = Path(getCarthageBuildPath(platform: target.platform))
+            var frameworkPath = platformPath + dependency.reference
+            if frameworkPath.extension == nil {
+                frameworkPath = Path(frameworkPath.string + ".framework")
+            }
```
[the remark this answers is not in the document]
Comment (brentleyjones) [reply Sources/XcodeGenKit/PBXProjGenerator.swift]: You can list extension as part of a `carthage` dependency. Also, both of these are just copying what was there in the above for loop. I would like to keep both of them the same for this PR, and if they should change they can in a different one. 😁

**Comment by brentleyjones** [comment]
@yonaskolb instead of a configurable value I could just change the default. Thinking more about it I think the default should only use the script for iOS Application targets. I’m going to be updating Carthage’s docs to say the same thing based on the conversation over there. Let me know if you want the value or not (I’ll update the default in a little bit).

**Comment by brentleyjones** [comment]
https://github.com/Carthage/Carthage/pull/2537 for update to their docs. Re-reading them they do say to only use it in iOS/tvOS/watchOS applications.

**Comment by brentleyjones** [comment]
You only need to strip when submitting to the App Store, so tests run on device are fine.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: Sources/XcodeGenKit/PBXProjGenerator.swift [hunk 2]
```diff
@@ -432,25 +432,28 @@ public class PBXProjGenerator {
         var copyResourcesReferences: [String] = []
         var copyWatchReferences: [String] = []
         var extensions: [String] = []
+        var carthageFrameworksToEmbed: [String] = []
 
         let targetDependencies = (target.transitivelyLinkDependencies ?? project.options.transitivelyLinkDependencies) ?
             getAllDependenciesPlusTransitiveNeedingEmbedding(target: target) : target.dependencies
+        
+        let directlyEmbedCarthage = target.directlyEmbedCarthageDependencies ?? !(target.platform.requiresSimulatorStripping && target.type.isApp)
```
[the remark this answers is not in the document]
Comment (brentleyjones) [reply Sources/XcodeGenKit/PBXProjGenerator.swift]: No. Tests have an extension of `.xctests`. `isApp` checks for `.app` correct?

[the remark this answers is not in the document]
Comment (yonaskolb) [reply Sources/XcodeGenKit/PBXProjGenerator.swift]: Check the getter, it includes test targets 👍 
`isApp` is used in other places. It's ok if you want to create new property and rename the usages to be more explicit about what they are checking

[the remark this answers is not in the document]
Comment (yonaskolb) [reply Sources/XcodeGenKit/PBXProjGenerator.swift]: Sorry I take all that back. I was thinking of `isExecutable`

**Comment by brentleyjones** [comment]
I agree on this being an option since you can't change it with a build setting.

**Pull request merged**